hcvault/keysource.go

```diff
@@ -0,0 +1,274 @@
+package hcvault
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/sirupsen/logrus"
+	"go.mozilla.org/sops/v3/logging"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("VAULT_TRANSIT")
+}
+
+// MasterKey is a Vault Transit backend path used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	EncryptedKey string
+	KeyName      string
+	EnginePath  string
+	VaultAddress string
+	CreationDate time.Time
+}
+
+// NewMasterKeysFromURIs gets lots of keys from lots of URIs
+func NewMasterKeysFromURIs(uris string) ([]*MasterKey, error) {
+	var keys []*MasterKey
+	if uris == "" {
+		return keys, nil
+	}
+	uriList := strings.Split(uris, ",")
+	for _, uri := range uriList {
+		if uri == "" {
+			continue
+		}
+		key, err := NewMasterKeyFromURI(uri)
+		if err != nil {
+			return nil, err
+		}
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+// NewMasterKeyFromURI obtains the vaultAddress the transit backend path and the key name from the full URI of the key
+func NewMasterKeyFromURI(uri string) (*MasterKey, error) {
+	log.Debugln("Called NewMasterKeyFromURI with uri: ", uri)
+	var key *MasterKey
+	if uri == "" {
+		return key, nil
+	}
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme == "" {
+		return nil, fmt.Errorf("missing scheme in vault URL (should be like this: https://vault.example.com:8200/v1/transit/keys/keyName), got: %v", uri)
+	}
+	enginePath, keyName, err := getBackendAndKeyFromPath(u.RequestURI())
+	if err != nil {
+		return nil, err
+	}
+	u.Path = ""
+	return NewMasterKey(u.String(), enginePath, keyName), nil
+
+}
+
+func getBackendAndKeyFromPath(fullPath string) (enginePath, keyName string, err error) {
+	// Running vault behind a reverse proxy with longer urls seems not to be supported
+	// by the vault client api so we have a separate Error for that here.
+	if re := regexp.MustCompile(`/[^/]+/v[\d]+/[^/]+/[^/]+/[^/]+`); re.Match([]byte(fullPath)) {
+		return "", "", fmt.Errorf("Running Vault with an prefixed url is not supported! (Format has to be like https://vault.example.com:8200/v1/transit/keys/keyName)")
+	} else if re := regexp.MustCompile(`/v[\d]+/[^/]+/[^/]+/[^/]+`); re.Match([]byte(fullPath)) == false {
+		return "", "", fmt.Errorf("Vault path does not seem to be formatted correctly: (eg. https://vault.example.com:8200/v1/transit/keys/keyName)")
```

Comment:
```suggestion
		return "", "", fmt.Errorf("vault path does not seem to be formatted correctly: (eg. https://vault.example.com:8200/v1/transit/keys/keyName)")
```

Go errors start with lowercase letters